haskell/test/Test/Plist/Property.hs

```diff
@@ -0,0 +1,140 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module Test.Plist.Property
+  ( hprop_Plist
+  ) where
+
+import Universum hiding (drop, swap)
+
+import qualified Data.List as DL
+import Data.List ((!!))
+import Fmt (build, unlinesF)
+import Hedgehog hiding (assert)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import Lorentz hiding (abs, assert, and, div, not, now, (>>))
+import Test.Cleveland
+
+import Ligo.BaseDAO.Types
+import SMT.Model.BaseDAO.Proposal.Plist
+import Test.Ligo.BaseDAO.Plist (genProposalKeyList)
+import Test.Plist.Type
+import Test.Plist.Contract
+
+
+genPlistParameter :: MonadGen m => [ProposalKey] -> [ProposalKey] -> m PlistParameter
+genPlistParameter keyList keyList2 = do
+  requireExistingKeyEps <-
+    if (not $ null keyList) then do
+      (i :: Int) <- Gen.integral (Range.constant 0 (length keyList - 1))
+      let k = keyList !! i
+      pure [ Mem k
+           , Delete k
+           ]
+    else pure []
+
+  notRequireExistingKeyEps <-
+    if (not $ null keyList2) then do
+      (i2 :: Int) <- Gen.integral (Range.constant 0 (length keyList2 - 1))
+      let k2 = keyList2 !! i2
+      pure [ Insert k2
+           , Mem k2
+           , Delete k2
+           ]
+    else pure []
+
+  Gen.choice $ pure <$> (requireExistingKeyEps <> notRequireExistingKeyEps <> [Pop ()])
+
+
+genPlistParameters :: MonadGen m => m ([ProposalKey], [PlistParameter])
+genPlistParameters = do
+  initSize <- Gen.integral (Range.constant 0 30)
+  initSize2 <- Gen.integral (Range.constant 0 30)
+
+  -- Init List to be used as storage
+  keyList <- genProposalKeyList initSize
+
+  -- A list to pick new key from
+  keyList2 <- genProposalKeyList initSize2
+
+  params <- Gen.list (Range.linear 10 20) $ genPlistParameter keyList keyList2
+
+  pure (keyList, params)
+
+
+hprop_Plist :: Property
+hprop_Plist =
+  withTests 100 $ property $ do
+    plistTests
+
+plistTests :: PropertyT IO ()
+plistTests = do
+  (keyList, params) <- forAll genPlistParameters
+  testScenarioProps $
+    scenarioEmulated $ do
+      let defaultStore = PlistStorage (plistFromList keyList) False
+      let haskellStore = (keyList, False)
+
+      -- Originate plist contract
+      plistContract <- originateTypedSimple @PlistParameter "PlistContract" defaultStore plistContractLigo
+
+      callContractLoop params (TAddress $ chAddress plistContract) haskellStore
+
+callContractLoop :: MonadEmulated caps base m => [PlistParameter] -> TAddress PlistParameter -> ([ProposalKey], Bool) -> m ()
+callContractLoop param addr haskellStore =
+  case param of
+    p:ps -> do
+      let newHaskellStore = callHaskell haskellStore p
+      newLigoStore <- callLigo addr p
+
+      assert (newHaskellStore == newLigoStore) $
```

Comment:
Oh, are we also comparing the result of `pop` calls somehow?

Reply:
I only compared the updated list, I've added the head result for the comparison as well.